kore-rpc-types/src/Kore/JsonRpc/Error.hs

```diff
@@ -0,0 +1,47 @@
+module Kore.JsonRpc.Error (module Kore.JsonRpc.Error) where
+
+import Control.Exception (ErrorCall (..), SomeException)
+import Control.Monad.Logger (logInfoN)
+import Data.Aeson
+import Data.Text qualified as Text
+import Kore.JsonRpc.Server (ErrorObj (..), JsonRpcHandler (..))
+import Text.Casing (fromHumps, toWords)
+
+-- RPC Server implementation errors
+
+cancelUnsupportedInBatchMode, notImplemented :: ErrorObj
+cancelUnsupportedInBatchMode = ErrorObj "Cancel request unsupported in batch mode" (-32001) Null
+notImplemented = ErrorObj "Not implemented" (-32002) Null
+
+unsupportedField :: Value -> ErrorObj
+unsupportedField = ErrorObj "Unsupported option" (-32003)
+
+-- Runtime backend errors
+
+data JsonRpcBackendError
+    = RuntimeError
+    | CouldNotParsePattern
+    | CouldNotVerifyPattern
+    | CouldNotFindModule
+    | ImplicationCheckError
+    | SmtSolverError
+    | Aborted
+    | MultipleStates
+    deriving stock (Enum, Show)
+
+backendError :: ToJSON a => JsonRpcBackendError -> a -> ErrorObj
```

Comment:
@tothtamas28 I've re-factored the rpc server errors into runtime errors caused by the haskell backend which start from 0, e.g. 

```json
{"code":4,"message":"Implication Check Error"}
```

vs server errors like the "Not implemented" error which is used by the booster for functionality currently not supported by it; these errors fall in the -32000 to -32099 range.


Let me know what you think

Reply:
Base on the name, I think it is OK to handle the following as "application defined errors":

```
CouldNotParsePattern
CouldNotVerifyPattern
CouldNotFindModule
ImplicationCheckError
SmtSolverError
MultipleStates
```

so in these cases, the error code indeed should be something outside of  the range `-32768..-32000`.

I'm not sure about `Aborted` and `RuntimeError`. When are these errors raised?

> server errors like the "Not implemented" error which is used by the booster for functionality currently not supported by it; these errors fall in the -32000 to -32099 range

I agree, this can qualify as an "implementation-defined server-error". (Unfortunately, the documentation is not very detailed about these labels.)

Reply:
`Aborted` is a bit of a blanket error when something has gone wrong sepcifically with the execution. `RuntimeError` is a blanket catch-all for any error raised in the backend (not the rpc server), this could be in execution/implication check/simplification/etc. This used to include the SMT error, and the idea is, if we find we need to pull out a runtime error/exception into a separate error-code we change that error from the blanket `RuntimeError` to a more specific one.

Reply:
> `Aborted` is a bit of a blanket error when something has gone wrong sepcifically with the execution

I agree this is an "application defined error" then, as I guess the client can debug the input / definition to work around the exception. Consider renaming it to `ExecutionError` or something similar.

> `RuntimeError` is a blanket catch-all for any error raised in the backend

So for example, if the server runs out of memory, it will raise `RutimeError`? Because that I'd classify as "implementation-defined server-error", as it's not the client's concern. For the same reason, I think having a single error code (from the range `-32000..-32099`) for all these conditions is fine.

Reply:
Hmm the line is hazy on `RuntimeError` because it can include stuff like IO errors when e.g. some resource wasn't found/is broken but also includes asserts statements in the code that have been violated, e.g. the recent `Cannot convert a denormalized substitution to a map!` from the slack channel would be thrown via this error too. I would thus suggest keeping it under application errors, because these errors will be coming in specifically from the application logic portion of the code

Reply:
My view is that "application error" is something that is meaningful to the client, i.e. the user of the application (e.g. "the SMT solver failed on your input"), whereas a "server error" is an internal error that should be kept hidden from the client (e.g. "OOM"). This is just an educated guess on my part based on analogies though, the JSON-RPC spec (or the XML-RPC spec it is derived from) does not elaborate on these terms.

But according to this logic, an assertion failure falls into the second category, as it indicates a bug in the code, and fixing that is beyond the client's control.

Also, if `RuntimeError` potentially stands for both application errors and server errors, I think the right approach is to be more restrictive, i.e. treat them as server errors by default.

But this is just my view of the issue, and I'm not strictly opposed to alterntives.

Reply:
Fair enough, I will move `RuntimeError` into the server error range and we can pull out any other errors into the application error scope as we see fit.